EventGenerator/src/CaloCalibGun_module.cc

```diff
@@ -0,0 +1,241 @@
+#include "Offline/MCDataProducts/inc/ProcessCode.hh"
+#include "Offline/CalorimeterGeom/inc/DiskCalorimeter.hh"
+#include "Offline/Mu2eUtilities/inc/RandomUnitSphere.hh"
+#include "Offline/MCDataProducts/inc/GenId.hh"
+#include "Offline/MCDataProducts/inc/GenParticle.hh"
+#include "Offline/MCDataProducts/inc/PrimaryParticle.hh"
+#include "Offline/MCDataProducts/inc/MCTrajectoryCollection.hh" 
+
+// Other external includes.
+#include "CLHEP/Random/RandFlat.h"
+#include "CLHEP/Vector/ThreeVector.h"
+#include "CLHEP/Vector/LorentzVector.h"
+#include "CLHEP/Random/RandomEngine.h"
+#include "CLHEP/Random/RandExponential.h"
+#include "CLHEP/Random/RandPoissonQ.h"
+
+#include "fhiclcpp/types/Atom.h"
+#include "messagefacility/MessageLogger/MessageLogger.h"
+
+using namespace std;
+
+namespace mu2e {
+
+  class CaloCalibGun : public art::EDProducer {
+  public:
+    struct Config {
+      using Name=fhicl::Name;
+      using Comment=fhicl::Comment;
+      fhicl::Atom<unsigned> mean{Name("mean"),1};
+      fhicl::Atom<double> energy{Name("PhotonEnergy"),6.13};//MeV
+      fhicl::Atom<double> cosmin{Name("cosmin"),-1.};
+      fhicl::Atom<double> cosmax{Name("cosmax"),1.};
+      fhicl::Atom<double> phimin{Name("phimin"),0.};
+      fhicl::Atom<double> phimax{Name("phimax"), CLHEP::twopi };
+      fhicl::Atom<double> tmin{Name("tmin"),0.};
+      fhicl::Atom<double> tmax{Name("tmax"),1694.};
+      fhicl::Atom<int> nDisk{Name("nDisk"),1};
+    };
+
+    using Parameters= art::EDProducer::Table<Config>;
+    explicit CaloCalibGun(const Parameters& conf);
+
+    virtual void produce(art::Event& event) override;
+
+    //----------------------------------------------------------------
+  private:
+
+    unsigned _mean;
+    double _energy;
+    double _cosmin;
+    double _cosmax;
+    double _phimin;
+    double _phimax;
+    double _tmin;
+    double _tmax;
+    int _nDisk;
+    
+    art::RandomNumberGenerator::base_engine_t& _engine;
+    CLHEP::RandFlat     _randFlat;
+    CLHEP::RandPoissonQ _randPoissonQ;
+    RandomUnitSphere    _randomUnitSphere;
+
+     //int                    _nPipes;
+    double                 _pipeRadius;
+    std::vector<double>    _pipeTorRadius;
+    std::vector<double>    _randomRad;
+    CLHEP::Hep3Vector      _zPipeCenter;
+
+  };
+
+  CaloCalibGun::CaloCalibGun(const Parameters& conf)
+    : EDProducer{conf}
+    , _mean{conf().mean()}
+    , _energy{conf().energy()}
+    , _cosmin{conf().cosmin()}
+    , _cosmax{conf().cosmax()}
+    , _phimin{conf().phimin()}
+    , _phimax{conf().phimax()}
+    , _tmin{conf().tmin()}
+    , _tmax{conf().tmax()}
+    , _nDisk{conf().nDisk()}
+    , _engine{createEngine(art::ServiceHandle<SeedService>()->getSeed())}
+    , _randFlat{_engine}
+    , _randPoissonQ{_engine, _mean*1.0}
+    , _randomUnitSphere{_engine, _cosmin, _cosmax, 0, CLHEP::twopi}
+  {
+    produces<mu2e::GenParticleCollection>();
+    produces<mu2e::PrimaryParticle>();
```

Comment:
I can't think why you need to make empty PrimaryParticle and MCTrajectoryData products?  If there is not a good reason, remove them.   If you get error messages without these, then something is misconfigured.

Reply:
Primary particle is necessary. The trajectory might not be

Reply:
Please explain what code uses a default constructed PrimaryParticle.